File: CSharpCodeBase/entities/player/character.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;
using System.Collections;
using MainGame.core;
//using UnityEngine.EventSystems;

namespace MainGame
{
    public class Character : UnityEntity
    {
        protected bool isDeath = false;
        protected float deathTimeout = 0f;
        protected bool isStatic = false;
        protected int characterClass;
        //characterClass = this.config:Get("characterClass")  ||  1;                        

        public void Awake()
        {
            //ConfigComponent config = ConfigComponent("heroes", name);        
            //gameObject.AddComponent<HealthComponnent>();
            //gameObject.AddComponent<CharacterDamageReceiver>();
            //gameObject.AddComponent<DamageVisualizerComponent>();
            //gameObject.AddComponent<MecanimComponent>();
            //gameObject.AddComponent<Mover>();
            //gameObject.AddComponent<CharacterWeaponContainer>();
            //gameObject.AddComponent<CombatComponent>();
            //gameObject.AddComponent<ResistComponent>();
            //gameObject.AddComponent<DamageProcessorComponent>();
            //gameObject.AddComponent<DamageReceiverComponent>();
            //gameObject.AddComponent<PlayerControllerComponent>();
            //gameObject.AddComponent<PossibleActionsComponent>();
            //gameObject.AddComponent<GrenageVisualizer>();
            //gameObject.AddComponent<GamePadMover>();
            //if(!GameController.inputService:IsGamepad()) {
            //  GetComponent<GamePadMover>().enabled = false;
            //}
            //gameObject.AddComponent<TimeManagerComponentr>();
            //GetComponent<Mover>().autoLook = false;            
        }                                                      
        
        public void Update()
        {           
        //// for left trigger
        //if self.gamepadRightStickController then
        //if GameController.inputService:LeftTriggerIsPressed() then
        //self.gamepadRightStickController.active = false
        //self.gamepadRightStickController:Disable()
        //else
        //self.gamepadRightStickController.active = true
        //end
        ///end  
        //  if(not this.isDeath  ){ 
        //    UnityExistsEntity.Update(self);
        //  }else{
        //    this.deathTimeout = this.deathTimeout - GameController.deltaTime;
        //    if(this.deathTimeout <= 0  ){ 
        //      self:Respawn();
        //    }
        //  }
        }
        public void FixedUpdate()
        {
            if (!isDeath)
                base.FixedUpdate();        
        }

        public void CanAttack(GameObject targetEntity)
        {
        // check weapon for(target entity
          //return this.weaponContainer:CanAttack(targetEntity);
        }
        
        public void Hit(object damageData)
        {
          //if damageData.summary != 0  ){ 
            SendMessage("Hit", damageData);
          //end
        }
        
        public void Pushed(object damageData){
        //if damageData.summary != 0  ){ 
            SendMessage("Pushed", damageData);
        //end
        }
        
        public void Death()
        {
            if (characterClass > 1)
            { 
                // mech
                //GetComponent<Mover>().Stop();
                isDeath = true;
                interactable = false;
                enabled = false;
                //GameController.player.playerController:DeleteCurrentSlot();
            }
            else
            {
                //GetComponent<Mecanim>():ForceSetState("Death");
                isDeath = true;
                interactable = false;
                deathTimeout = 5;
                //GetComponent<Mover>:Stop();
          }
        }
        
        public void Respawn()
        {
            SendMessage("OnRespawn");
            //SetPosition(GameController.player.homePosition) ;
            //GameController.worldController:TeleportTargetTransform(this.transform, ;
            //GameController.player.homePosition.x,;
            //GameController.player.homePosition.y,;
            //GameController.player.homePosition.z);
            //GetComponent<Mecanim>():ForceSetState("Idle");
            interactable = true;
            isDeath = false;
        }
        
        public void OnSelectCharacter()
        {
            //  GameController.ui.SetupIcon(this.config:Get("cpanel_icon"));
            //  GameController.ui.SetupAmount(this.health:GetPercentAmount());
            //  //GameController.ui.UpdateHealth(this.health:GetPercentAmount())
            Debug.Log("Spell system initialization!");
            //  GameController.spellSystem:SetupSpells(this.config:Get("spell"));
            //  //GameController.ui.slotsController:SetContainerSubTag("weapon", this.config:Get("weaponTag"))  
            interactable = true;
            enabled = true;
        }

        public void OnDeselectCharacter()
        {
            //GameController.spellSystem:SaveStatuses();
            interactable = false;
            this.enabled = false;
            //if(this.gamepadRightStickController  ){ 
            //this.gamepadRightStickController:DropTarget();
            //}  
        }

        /* FiX ME AFTER SPELL will be defined
        public void OnSpellCast(Spell spell)
        {
        //  return this.playerController:OnSpellCast(spell);
        }
         */

        public void OnHealthChanged()
        {
        //  GameController.ui.SetupAmount(this.health:GetPercentAmount());
        //  //GameController.ui.UpdateHealth(this.health:GetPercentAmount())
        }

        public void OnCollisionEnter(GameObject targetEntity)
        {
        }

        public void SetTimer(string name, float value)  {
            //GetComponent<TimeManager>():Add("TimerHandler", value, name, "extended");
        }
        /*
        public void TimerHandler(tResult)    {
        //  if(tResult.timeLeft <= 0  ){ 
        //    tResult.complete = true;
        //    GameController.spellSystem:TimerUpdate(this.name, tResult.name, 0)    ;
        //  }else{
        //    GameController.spellSystem:TimerUpdate(this.name, tResult.name, tResult.timeLeft);
        //  }  
        }
         */        
    }
}

File: CSharpCodeBase/entities/player/character.lua
local Class = require 'utils.hump.class'
local UnityExistsEntity = require 'unity.unityentity'
local Vector3 = require 'utils.hump.vector3'

local CharacterWeaponContainer = require 'components.character.weaponcontainer'
local Mover = require 'components.character.mover'
local MecanimComponent = require 'components.animations.mecanim'
local PlayerControllerComponent = require 'components.controlsystem.playercontroller'
local CombatComponent = require 'components.character.combat'
local ResistComponent = require 'components.character.resist'
local DamageProcessorComponent = require 'components.character.damageprocessor'
local DamageReceiverComponent = require 'components.character.damagereceiver'
local ConfigComponent = require 'components.services.config'
local PossibleActionsComponent = require 'components.character.possibleactions'

local DamageVisualizerComponent = require 'components.character.damagevisualizer'
local HealthComponent = require 'components.character.health'
local CharacterDamageReceiver = require 'components.character.characterdamagereceiver'
local GrenageVisualizer = require 'components.character.grenademodevisualizer'
local GamePadMover = require 'components.character.gamepadmover'
local TimeManagerComponent = require 'components.thing.timemanagercomponent'

local CharacterEntity = Class({__includes=UnityExistsEntity, init = function (self, gameObject)
      UnityExistsEntity.init(self, gameObject)
      self.config = ConfigComponent('heroes', self.name)
      self.rigidbody = self.gameObject:GetComponent("Rigidbody")
      self:AddComponent(HealthComponent)
      self:AddComponent(CharacterDamageReceiver)
      self:AddComponent(DamageVisualizerComponent)
      self:AddComponent(MecanimComponent)      
      self:AddComponent(Mover)        
      self:AddComponent(CharacterWeaponContainer)      
      self:AddComponent(CombatComponent)
      self:AddComponent(ResistComponent)
      self:AddComponent(DamageProcessorComponent)
      self:AddComponent(DamageReceiverComponent)
      self:AddComponent(PlayerControllerComponent)
      self:AddComponent(PossibleActionsComponent)
      self:AddComponent(GrenageVisualizer)
      if GameController.inputService:IsGamepad() then
        self:AddComponent(GamePadMover)
      end
      self:AddComponent(TimeManagerComponent)
      self.luaMapper:SetAlwaysVisible(true)
      self.isDeath = false
      self.deathTimeout = 0
      self.characterClass = self.config:Get('characterClass') or 1
      self.static = false
      self.mover.autoLook = false
end})

function CharacterEntity:Update()
  --[[ for left trigger
  if self.gamepadRightStickController then
    if GameController.inputService:LeftTriggerIsPressed() then
      self.gamepadRightStickController.active = false
      self.gamepadRightStickController:Disable()
    else
      self.gamepadRightStickController.active = true
    end
  end
  ]]  
  if not self.isDeath then
    UnityExistsEntity.Update(self)
  else
    self.deathTimeout = self.deathTimeout - GameController.deltaTime
    if self.deathTimeout <= 0 then
      self:Respawn()
    end
  end
end

function CharacterEntity:FixedUpdate()
  if not self.isDeath then
    UnityExistsEntity.FixedUpdate(self)
  end
end

function CharacterEntity:CanAttack(targetEntity)
  -- check weapon for target entity
  return self.weaponContainer:CanAttack(targetEntity)
end

function CharacterEntity:OnEvent(data)
  UnityExistsEntity.OnEvent(self, data)
end

function CharacterEntity:Hit(damageData)
  --if damageData.summary ~= 0 then
    self:Message("Hit", damageData)
  --end
end

function CharacterEntity:Pushed(damageData)
  --if damageData.summary ~= 0 then
    self:Message("Pushed", damageData)
  --end
end

function CharacterEntity:Death()
  if self.characterClass > 1 then
    -- mech
    self.mover:Stop()
    self.isDeath = true
    self.interactable = false
    self.enabled = false
    GameController.player.playerController:DeleteCurrentSlot()
  else
    self.mecanim:ForceSetState('Death')
    self.isDeath = true
    self.interactable = false
    self.deathTimeout = 5
    self.mover:Stop()
  end
end

function CharacterEntity:Respawn()
  self:Message('OnRespawn')
  self:SetPosition(GameController.player.homePosition) 
  GameController.worldController:TeleportTargetTransform(self.transform, 
    GameController.player.homePosition.x,
    GameController.player.homePosition.y,
    GameController.player.homePosition.z)
  self.mecanim:ForceSetState('Idle')
  self.interactable = true
  self.isDeath = false
end

function CharacterEntity:OnSelectCharacter()
  GameController.ui.character:SetupIcon(self.config:Get('cpanel_icon'))
  GameController.ui.character:SetupAmount(self.health:GetPercentAmount())
  --GameController.ui.character:UpdateHealth(self.health:GetPercentAmount())
  print("Spell system initialization!")
  GameController.spellSystem:SetupSpells(self.config:Get('spell'))
  --GameController.ui.slotsController:SetContainerSubTag('weapon', self.config:Get('weaponTag'))  
  self.interactable = true
  self.enabled = true
end

function CharacterEntity:OnDeselectCharacter()  
  GameController.spellSystem:SaveStatuses()
  self.interactable = false
  self.enabled = false
  if self.gamepadRightStickController then
    self.gamepadRightStickController:DropTarget()
  end  
end

function CharacterEntity:OnSpellCast(spell)
  return self.playerController:OnSpellCast(spell)
end

function CharacterEntity:OnHealthChanged()
  GameController.ui.character:SetupAmount(self.health:GetPercentAmount())
  --GameController.ui.character:UpdateHealth(self.health:GetPercentAmount())
end

function CharacterEntity:OnCollisionEnter(targetEntity)
end

function CharacterEntity:SetTimer(name, value)  
  self.timeManager:Add("TimerHandler", value, name, 'extended')
end

function CharacterEntity:TimerHandler(tResult)    
  if tResult.timeLeft <= 0 then
    tResult.complete = true
    GameController.spellSystem:TimerUpdate(self.name, tResult.name, 0)    
  else
    GameController.spellSystem:TimerUpdate(self.name, tResult.name, tResult.timeLeft)
  end  
end

return CharacterEntity